[16.0][IMP] cetmix_tower_server/yaml UI/UX updates

ivs-cetmix opened the pull request:
Forward port updates from 14.0 

## Server https://github.com/cetmix/cetmix-tower/pull/257 https://github.com/cetmix/cetmix-tower/pull/267  https://github.com/cetmix/cetmix-tower/pull/276

-  New 'Partners' submenu in 'Servers' menu.
- 'Cetmix Tower' tab in partner form showing all secret keys of this partner.
- 'Servers' button in partner form with that shows the number of servers of this partner.
- Search server by partner name or IP address.
- Show access level in the command and flight plan list views.
- Hide applicability selector for members of the 'User' group in the command and flight plan run wizards.
- Update demo date to improve user experience when testing command and plans as the "Demo" user.

## Yaml https://github.com/cetmix/cetmix-tower/pull/277

Move groups to separate categories and rename them to "Allow" in order to
make them visible without enabling the developer mode.

Task: 4642


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Added server and secret key management features to partner records, including server count display and related actions.
  - Introduced a "Partners" menu and list to access partner-server relationships.
  - Added UI/UX improvements and new group-by filters for commands and plans by access level.
  - User groups are now visible without developer mode.

- **Enhancements**
  - Improved server search views with additional searchable fields.
  - Added context defaults for SSH key selection in server templates.
  - Updated demo data to explicitly associate servers and access levels.

- **Access Control**
  - Refined group-based access to command and plan execution wizards and related fields.
  - Enhanced access control for general settings and form elements based on user roles.

- **Bug Fixes**
  - Corrected server count computation across partner hierarchies.
  - Fixed default applicability and server visibility logic in command and plan wizards.

- **Documentation**
  - Updated documentation and badges to use British English spelling "licence".
  - Removed obsolete news fragments and clarified user-facing notes.

- **Tests**
  - Added tests covering partner-server relationships and wizard applicability behavior.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Mohajiro committed:
[IMP] cetmix_tower_server: UI/UX updates

-  New 'Partners' submenu in 'Servers' menu.
- 'Cetmix Tower' tab in partner form showing all secret keys of this partner.
- 'Servers' button in partner form with that shows the number of servers
  of this partner.
- Search server by partner name or IP address.
- Show access level in the command and flight plan list views.
- Hide applicability selector for members of the 'User' group in the
  command and flight plan run wizards.

Task: 4642

ivs-cetmix committed:
[IMP] cetmix_tower_yaml: Rearrange groups

Move groups to separate categories and rename them to "Allow" in order to
make them visible without enabling the developer mode.

Task: 4638

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/cetmix/cetmix-tower?pullRequest=278) <br/>All committers have signed the CLA.

ivs-cetmix commented on cetmix_tower_server/models/__init__.py:
```diff
@@ -28,3 +28,4 @@
 from . import cx_tower_variable_option
 from . import ir_actions_server
 from . import res_config_settings
+from . import res_partner
```
This is fine, let's leave it as is

ivs-cetmix commented on cetmix_tower/readme/newsfragments/4642.feature:
```diff
@@ -0,0 +1 @@
+UI/UX improvements.
```
This is fine, let's leave it as is

ivs-cetmix commented on cetmix_tower_server/views/menuitems.xml:
```diff
@@ -31,6 +31,15 @@
         groups="group_manager"
         sequence="10"
     />
+    <menuitem
+        id="menu_cx_tower_partner"
+        name="Partners"
+        action="action_cx_cetmix_tower_partner"
+        parent="menu_cx_tower_server_root"
+        groups="cetmix_tower_server.group_user"
+        sequence="15"
+    />
```
Noted, thank you!

ivs-cetmix commented on cetmix_tower_server/data/ir_actions_server.xml:
```diff
@@ -23,4 +23,11 @@
         <field name="groups_id" eval="[(4, ref('cetmix_tower_server.group_user'))]" />
     </record>
 
+    <record id="action_cx_cetmix_tower_partner" model="ir.actions.act_window">
+      <field name="name">Partners</field>
+      <field name="type">ir.actions.act_window</field>
+      <field name="res_model">res.partner</field>
+      <field name="view_mode">tree,form</field>
+    </record>
```
This is not needed because we would like to show all partners

ivs-cetmix commented on cetmix_tower_server/README.rst:
```diff
@@ -13,7 +13,7 @@ Cetmix Tower Server
-.. |badge2| image:: https://img.shields.io/badge/license-AGPL--3-blue.png
+.. |badge2| image:: https://img.shields.io/badge/licence-AGPL--3-blue.png
     :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
     :alt: License: AGPL-3
```
This file is auto generated. How many times should I repeat this for you to remember?

ivs-cetmix commented on cetmix_tower_yaml/README.rst:
```diff
@@ -13,7 +13,7 @@ Cetmix Tower YAML
-.. |badge2| image:: https://img.shields.io/badge/license-AGPL--3-blue.png
+.. |badge2| image:: https://img.shields.io/badge/licence-AGPL--3-blue.png
     :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
     :alt: License: AGPL-3
```
This file is auto generated. How many times should I repeat this for you to remember?

ivs-cetmix commented on cetmix_tower_server/static/description/index.html:
```diff
@@ -369,7 +369,7 @@ <h1 class="title">Cetmix Tower Server</h1>
 !! source digest: sha256:8123f8ca0290bb006dcaaf5467c4f8cba25ed83bb1949f435a9cf4f688989aa1
 !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! -->
-<p><a class="reference external image-reference" href="https://odoo-community.org/page/development-status"><img alt="Beta" src="https://img.shields.io/badge/maturity-Beta-yellow.png" /></a> <a class="reference external image-reference" href="http://www.gnu.org/licenses/agpl-3.0-standalone.html"><img alt="License: AGPL-3" src="https://img.shields.io/badge/license-AGPL--3-blue.png" /></a> <a class= ...
+<p><a class="reference external image-reference" href="https://odoo-community.org/page/development-status"><img alt="Beta" src="https://img.shields.io/badge/maturity-Beta-yellow.png" /></a> <a class="reference external image-reference" href="http://www.gnu.org/licenses/agpl-3.0-standalone.html"><img alt="License: AGPL-3" src="https://img.shields.io/badge/licence-AGPL--3-blue.png" /></a> <a class= ...
```
This file is auto generated. How many times should I repeat this for you to remember?

ivs-cetmix commented on cetmix_tower_yaml/static/description/index.html:
```diff
@@ -369,7 +369,7 @@ <h1 class="title">Cetmix Tower YAML</h1>
 !! source digest: sha256:676fd1a19d947bff362bdeb0e64dae50cb784227d063b9a819de7996fbefd05b
 !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! -->
-<p><a class="reference external image-reference" href="https://odoo-community.org/page/development-status"><img alt="Beta" src="https://img.shields.io/badge/maturity-Beta-yellow.png" /></a> <a class="reference external image-reference" href="http://www.gnu.org/licenses/agpl-3.0-standalone.html"><img alt="License: AGPL-3" src="https://img.shields.io/badge/license-AGPL--3-blue.png" /></a> <a class= ...
+<p><a class="reference external image-reference" href="https://odoo-community.org/page/development-status"><img alt="Beta" src="https://img.shields.io/badge/maturity-Beta-yellow.png" /></a> <a class="reference external image-reference" href="http://www.gnu.org/licenses/agpl-3.0-standalone.html"><img alt="License: AGPL-3" src="https://img.shields.io/badge/licence-AGPL--3-blue.png" /></a> <a class= ...
```
This file is auto generated. How many times should I repeat this for you to remember?

GabbasovDinar commented on cetmix_tower_server/wizards/cx_tower_command_run_wizard.py:
```diff
@@ -94,20 +94,25 @@ class CxTowerCommandRunWizard(models.TransientModel):
-            )
+        for rec in self:
+            rec.show_servers = bool(rec.server_ids and len(rec.server_ids) > 1)
+            if not self._is_privileged_user():
```
we can move it before cycle e.g.:
```python

is_privileged_user = self._is_privileged_user()
for rec in self:
    ...
```

GabbasovDinar replied:
and some refactoring:

```python
is_privileged_user = self._is_privileged_user()
for rec in self:
    vals = {
        "show_servers":  bool(rec.server_ids and len(rec.server_ids) > 1)
    }
    if not is_privileged_user:
        vals["applicability"] = "this"
    rec.update(vals)
```

ivs-cetmix replied:
This is an obsolete method, will remove it completely. Thank you for noticing this!

ivs-cetmix committed:
[IMP] cetmix_tower_server: Demo data

Update demo data to include access level and server ids.
This is needed to improve user experience when testing flight plans and
commands as a "Demo" user.

Task: 4642

ivs-cetmix commented on cetmix_tower_server/views/cx_tower_server_view.xml:
```diff
@@ -452,9 +452,13 @@
                 <field name="reference" />
-                <field name="status" />
-                <field name="os_id" />
+                <field name="url" />
+                <field name="partner_id" />
                 <field name="tag_ids" />
+                <field name="ip_v4_address" />
+                <field name="ip_v6_address" />
+                <field name="os_id" />
+                <field name="status" />
```
This is a search view. It doesn't support 'optional' attribute.

ivs-cetmix commented on cetmix_tower_yaml/security/cetmix_tower_yaml_groups.xml:
```diff
@@ -1,22 +1,27 @@
 <record id="group_export" model="res.groups">
-    <field name="name">Export</field>
-    <field name="category_id" ref="ir_module_category_tower_yaml" />
+    <field name="name">Allow</field>
+    <field name="category_id" ref="ir_module_category_tower_yaml_export" />
     <field name="comment">
```
Group label is "YAML Export". And "Allow" is a value. So final expressions looks like "YAML Export: Allow"

ivs-cetmix commented on cetmix_tower_server/wizards/cx_tower_command_run_wizard.py:
```diff
@@ -419,6 +416,12 @@ def run_command_in_wizard(self):
                 "target": "new",
             }
 
+    def _is_privileged_user(self):
+        """Return True if current user is in Manager or Root group."""
+        return self.env.user.has_group(
+            "cetmix_tower_server.group_manager"
+        ) or self.env.user.has_group("cetmix_tower_server.group_root")
+
```
Thank you for suggestion, will keep an eye on it in the future.

ivs-cetmix commented on cetmix_tower_server/tests/test_partner_server_btn.py:
```diff
@@ -0,0 +1,58 @@
+            "skip_host_key": True,
+        }
+
+        cls.Server.create({"partner_id": cls.partner_b.id, **cls.server_defaults})
+        cls.Server.create({"partner_id": cls.partner_b.id, **cls.server_defaults})
+        cls.Server.create({"partner_id": cls.partner_b_child.id, **cls.server_defaults})
```
Let's leave it as is for now.

ivs-cetmix commented on cetmix_tower_server/tests/test_partner_server_btn.py:
```diff
@@ -0,0 +1,58 @@
+        cls.Server.create({"partner_id": cls.partner_b.id, **cls.server_defaults})
+        cls.Server.create({"partner_id": cls.partner_b_child.id, **cls.server_defaults})
+
+        key = cls.Key.create({"name": "SSH Token", "key_type": "s"})
+        cls.KeyValue.create(
+            {
+                "key_id": key.id,
+                "partner_id": cls.partner_b.id,
+                "secret_value": "TOPSECRET",
+            }
+        )
```
Let's leave it as is for now.

ivs-cetmix commented on cetmix_tower_server/readme/newsfragments/4642.feature:
```diff
@@ -0,0 +1 @@
+UI/UX improvements.
```
This is enough for a newsfragment

ivs-cetmix commented on cetmix_tower_yaml/security/cetmix_tower_yaml_groups.xml:
```diff
@@ -1,22 +1,27 @@
 <record id="group_import" model="res.groups">
-    <field name="name">Import</field>
-    <field name="category_id" ref="ir_module_category_tower_yaml" />
+    <field name="name">Allow</field>
+    <field name="category_id" ref="ir_module_category_tower_yaml_import" />
     <field name="comment">
```
This is already explained in the comment below

ivs-cetmix commented on cetmix_tower_server/views/cx_tower_server_view.xml:
```diff
@@ -402,14 +402,11 @@
-                                context="{'default_key_type': 's', 'from_server': True}"
-                            >
+                            <field name="secret_ids">
                                 <tree editable="bottom">
                                     <field
                                         name="key_id"
-                                        domain="[('key_type', '=', 's')]"
+                                        context="{'default_key_type': 's', 'secrets_only': True}"
                                     />
```
How about this? https://github.com/cetmix/cetmix-tower/blob/5f80f284d899bc8ff2c82e64606553a944acb7cf/cetmix_tower_server/views/cx_tower_server_view.xml#L409

ivs-cetmix commented on cetmix_tower_server/tests/test_partner_server_btn.py:
```diff
@@ -0,0 +1,58 @@
+            }
+        )
+
+        cls.server_defaults = {
+            "name": "Test Server",
+            "ssh_username": "root",
+            "ssh_port": 22,
+            "ssh_password": "Test-P@ssw0rd-123",
+            "ip_v4_address": "127.0.0.1",
+            "skip_host_key": True,
+        }
```
Let's leave it as is

ivs-cetmix commented on cetmix_tower_server/tests/test_partner_server_btn.py:
```diff
@@ -0,0 +1,58 @@
+        )
+
+    def test_server_count_compute(self):
+        """Server count: direct + one‑level child + zero if none."""
+        self.assertEqual(self.partner_b.server_count, 3)
```
Looks like this is the same symbol

Aldeigja reviewed (approved):
Functional LGTM

ivs-cetmix commented:
/ocabot merge patch

CetmixGitDrone commented:
On my way to merge this fine PR!
Prepared branch [16.0-ocabot-merge-pr-278-by-ivs-cetmix-bump-patch](https://github.com/cetmix/cetmix-tower/commits/16.0-ocabot-merge-pr-278-by-ivs-cetmix-bump-patch), awaiting test results.

CetmixGitDrone commented:
Congratulations, your PR was merged at 65a9af743f60b6711ef3d232a641544c6f46609e. Thanks a lot for contributing to cetmix. ❤️
